engine/instant_query.go

```diff
@@ -15,32 +17,82 @@ import (
 
 type instantQuery struct {
 	plan model.VectorOperator
+	pool *model.VectorPool
+	expr parser.Expr
+	ts   time.Time
 }
 
-func newInstantQuery(plan model.VectorOperator) promql.Query {
-	return &instantQuery{plan: plan}
+func newInstantQuery(plan model.VectorOperator, pool *model.VectorPool, expr parser.Expr, ts time.Time) promql.Query {
+	return &instantQuery{
+		plan: plan,
+		pool: pool,
+		expr: expr,
+		ts:   ts,
+	}
 }
 
 func (q *instantQuery) Exec(ctx context.Context) *promql.Result {
-	return nil
+	vs, err := q.plan.Next(ctx)
+	if err != nil {
+		return newErrResult(err)
+	}
+
+	if len(vs) == 0 {
+		return &promql.Result{}
+	}
+
+	resultSeries, err := q.plan.Series(ctx)
+	if err != nil {
+		return newErrResult(err)
+	}
+
+	series := make([]promql.Series, len(resultSeries))
+	for i := 0; i < len(resultSeries); i++ {
+		series[i].Metric = resultSeries[i]
+	}
+
+	for _, vector := range vs {
+		for _, sample := range vector.Samples {
+			if len(series[sample.ID].Points) == 0 {
+				series[sample.ID].Points = make([]promql.Point, 0, 121)
+			}
+			series[sample.ID].Points = append(series[sample.ID].Points, promql.Point{
+				T: vector.T,
+				V: sample.V,
+			})
+			q.plan.GetPool().PutSamples(vector.Samples)
+		}
+	}
+	q.plan.GetPool().PutVectors(vs)
+
+	var result parser.Value
+	switch q.expr.Type() {
+	case parser.ValueTypeMatrix:
+		result = promql.Matrix(series)
+	case parser.ValueTypeVector:
+		// Convert matrix with one value per series into vector.
+		vector := make(promql.Vector, len(resultSeries))
+		for i := range series {
+			// Point might have a different timestamp, force it to the evaluation
+			// timestamp as that is when we ran the evaluation.
+			vector[i] = promql.Sample{
+				Metric: series[i].Metric,
+				Point: promql.Point{
+					V: series[i].Points[0].V,
+					T: q.ts.UnixMilli(),
+				},
+			}
+		}
+		result = vector
+	case parser.ValueTypeString:
+		result = promql.Scalar{V: series[0].Points[0].V, T: q.ts.UnixMilli()}
```

Comment:
This should be `promql.String`?
And we also need to handle the scalar case with `parser.ValueTypeScalar`

Reply:
Nice catch! I was planning to handle the parser.ValueTypeScalar here. Handling string might need a small change in plan.go. Will implement! 🙂